swaybar/i3bar.c

```diff
@@ -0,0 +1,216 @@
+#define _POSIX_C_SOURCE 200809L
+#include <json-c/json.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <wlr/util/log.h>
+#include "swaybar/config.h"
+#include "swaybar/status_line.h"
+
+static void i3bar_block_free(struct i3bar_block *block) {
+	if (!block) {
+		return;
+	}
+	wl_list_remove(&block->link);
+	free(block->full_text);
+	free(block->short_text);
+	free(block->align);
+	free(block->name);
+	free(block->instance);
+	free(block->color);
+}
+
+static bool i3bar_parse_json(struct status_line *status, const char *text) {
+	struct i3bar_block *block, *tmp;
+	wl_list_for_each_safe(block, tmp, &status->blocks, link) {
+		i3bar_block_free(block);
+	}
+	json_object *results = json_tokener_parse(text);
+	if (!results) {
+		status_error(status, "[failed to parse i3bar json]");
+		return false;
+	}
+	if (json_object_array_length(results) < 1) {
+		return true;
+	}
+	for (size_t i = 0; i < json_object_array_length(results); ++i) {
+		json_object *full_text, *short_text, *color, *min_width, *align, *urgent;
+		json_object *name, *instance, *separator, *separator_block_width;
+		json_object *background, *border, *border_top, *border_bottom;
+		json_object *border_left, *border_right, *markup;
+		json_object *json = json_object_array_get_idx(results, i);
+		if (!json) {
+			continue;
+		}
+		json_object_object_get_ex(json, "full_text", &full_text);
+		json_object_object_get_ex(json, "short_text", &short_text);
+		json_object_object_get_ex(json, "color", &color);
+		json_object_object_get_ex(json, "min_width", &min_width);
+		json_object_object_get_ex(json, "align", &align);
+		json_object_object_get_ex(json, "urgent", &urgent);
+		json_object_object_get_ex(json, "name", &name);
+		json_object_object_get_ex(json, "instance", &instance);
+		json_object_object_get_ex(json, "markup", &markup);
+		json_object_object_get_ex(json, "separator", &separator);
+		json_object_object_get_ex(json, "separator_block_width", &separator_block_width);
+		json_object_object_get_ex(json, "background", &background);
+		json_object_object_get_ex(json, "border", &border);
+		json_object_object_get_ex(json, "border_top", &border_top);
+		json_object_object_get_ex(json, "border_bottom", &border_bottom);
+		json_object_object_get_ex(json, "border_left", &border_left);
+		json_object_object_get_ex(json, "border_right", &border_right);
+
+		struct i3bar_block *block = calloc(1, sizeof(struct i3bar_block));
+		block->full_text = full_text ?
+			strdup(json_object_get_string(full_text)) : NULL;
+		block->short_text = short_text ?
+			strdup(json_object_get_string(short_text)) : NULL;
+		if (color) {
+			block->color = malloc(sizeof(uint32_t));
```

Comment:
Allocating a single uint32_t?

Reply:
Yeah cause it can be NULL

Reply:
0 != NULL in this case

Reply:
I think an accompanying "is_valid" bool would be more appropriate.

Reply:
nack